### ISSUE TITLE
vg giraffe and HPRC graphs running slow

### ISSUE THREAD

**Issue by hannespetur** [vgteam/vg#3899]
Hello,

vg giraffe is running quite a bit slower than expected on the HPRC graphs based on the reported numbers. Can you help me debug this problem?

**1. What were you trying to do?**
I would like to run vg giraffe + deepvariant on the HPRC graph. Similar to what is described in:
https://www.biorxiv.org/content/10.1101/2022.10.06.511217v2.full.pdf

To do this I am using this wdl:
https://github.com/vgteam/vg_wdl/blob/1c24d234d1f760bdafd6565b50d917eb5ae44046/workflows/giraffe_and_deepvariant.wdl

with input data from:
https://s3-us-west-2.amazonaws.com/human-pangenomics/index.html?prefix=pangenomes/freeze/freeze1/minigraph-cactus/filtered/

The `*grch38-maxdel.10mb*` files

**2. What did you want to happen?**
The runVGGIRAFFE task of the pipeline to run at mapping speed similar to the reported numbers in the Minigraph-Cactus pipeline (1800-2500 reads/thread/second). 

**3. What actually happened?**
My runs have mapping speed around 5 reads/thread/second or 400x slower than the reported speed. The problem is only the speed, the output itself looks good to me.

**4. If you got a line like `Stack trace path: /somewhere/on/your/computer/stacktrace.txt`, please copy-paste the contents of that file here:**

Not stacktrace but here is an example log message from vg giraffe

```
[IndexRegistry]: Combining Giraffe GBWT and GBWTGraph into GBZ.
Loading Distance Index v2
Initializing MinimizerMapper
Loading and initialization: 944.195 seconds
Mapping reads to "-" (GAM)
--hit-cap 10
--hard-hit-cap 500
--score-fraction 0.9
--max-min 500
--num-bp-per-min 1000
--exclude-overlapping-min
--max-extensions 800
--max-alignments 8
--cluster-score 50
--pad-cluster-score 0
--cluster-coverage 0.3
--extension-score 1
--extension-set 20
--max-multimaps 1
--distance-limit 200
--paired-distance-limit 2
--rescue-subgraph-size 4
--rescue-seed-limit 100
--rescue-attempts 15
--rescue-algorithm dozeu
warning[vg::giraffe]: Refusing to perform too-large rescue alignment of 151 bp against 14290 bp ordered subgraph for read A00744:46:HV3C3DSXX:2:2671:24234:24173 which would use more than 1572864 cells and might exhaust Dozeu's allocator; suppressing further warnings.
Using fragment length estimate: 407.406 +/- 168.155
warning[vg::giraffe]: Refusing to perform too-large tail alignment of 108 bp against 16552 bp tree which would use more than 1572864 cells and might exhaust Dozeu's allocator; suppressing further warnings.
warning: trying to find the distance in an oversized snarl without a graph. Returning inf
Mapped 4000000 reads across 15 threads in 50069.5 seconds with 35.8604 additional single-threaded seconds.
Mapping speed: 5.32568 reads per second per thread
Used 554159 CPU-seconds (including output).
Achieved 7.21815 reads per CPU-second (including output)
Used 1955035493810942 CPU instructions (not including output).
Mapping slowness: 488.759 M instructions per read at 3527.93 M mapping instructions per inclusive CPU-second
Memory footprint: 56.4875 GB
```

**5. What data and command can the vg dev team use to make the problem happen?**

See wdl and input data above.

**6. What does running `vg version` say?**

The problem is both produced with vg v1.37.0 (the one used in the minigraph-cactus paper) and vg v1.44.0

```
vg version v1.37.0 "Monchio"
Compiled with g++ (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0 on Linux
Linked against libstd++ 20200808
Built by root@bbcc264db88d

vg version v1.44.0 "Solara"
Compiled with g++ (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0 on Linux
Linked against libstd++ 20210601
Built by root@buildkitsandbox
```

Best,
Hannes

**Comment by glennhickey** [comment]
We did all our mapping experiments with the allele-frequency filtered graphs, `*minaf.0.1*`.  This may be the issue.  If so, I didn't realize the speed difference was so stark, and will have to make the description on the website more clear.  Could you please try on the filtered graph?

**Comment by hannespetur** [comment]
Thanks for the reply. Sure no problem, I will try that and report back how it goes. But my understanding was that the ~1800 reads/thread/second (column 0 on Supplementary Figure 9 in the MC preprint) was on the unfiltered graph, so I was expecting to get similar or faster speed with the `maxdel.10mb` filtered graph.

Best,
Hannes

**Comment by hannespetur** [comment]
Thanks for digging up these details for me! I don't mind starting with the CHM13 graph, so I will instead try to use the exact same parameters.

**Comment by glennhickey** [comment]
OK, but there should not be any appreciable difference in speed between the grch38-based and chm13-based graph.  But if you want to reproduce that data point, that's how to do it! 
For what it's worth I think the link to the exact workflow is this one https://github.com/vgteam/vg_wdl/blob/giraffedv-v1.3-mc/workflows/giraffe_and_deepvariant.wdl (which is linked off the dockstore page I sent above).

**Comment by hannespetur** [comment]
The mapping speed got a lot better!

With exact same parameters as you listed above, vg reports these metrics:
`hprc-v1.0-mc-chm13` 1150 reads/thread/second
`hprc-v1.0-mc-grch38-minaf.0.1` 2600 reads/thread/second

I am using the exact same graph and FASTQs as before. The workflow has very small changes which seem irrelevant to my problem so I can only image it is the vg version.

Thanks for all the help Glenn!

Best,
Hannes

**Comment by glennhickey** [comment]
Wow, it still seems like there is something wrong. For one, there shouldn't be such a difference between those graphs.  For two, the commit used in my script should correspond pretty closely to v.1.44.0.  But you're saying this difference in versions is the reason?

**Comment by glennhickey** [comment]
I've got a newly-indexed hprc graph I'm mapping some reads to.  It shouldn't be terribly different than the usual hprc graphs but does have a bunch of reference path fragments.  

So far giraffe (vg 1.47) has mapped 15178 reads in 72840 cpu seconds.  stderr has 346 warnings like 
```
warning[vg::Watchdog]: Thread 66 finally checked out after 132 seconds and 0 kb memory growth processing: A00744:46:HV3C3DSXX:2:1101:15239:2237, A00744:46:HV3C3DSXX:2:1101:15239:2237
warning[vg::Watchdog]: Thread 60 has been checked in for 10 seconds processing: A00744:46:HV3C3DSXX:2:1101:10800:7795, A00744:46:HV3C3DSXX:2:1101:10800:7795
warning[vg::Watchdog]: Thread 69 finally checked out after 691 seconds and 1193152 kb memory growth processing: A00744:46:HV3C3DSXX:2:1101:25681:1407, A00744:46:HV3C3DSXX:2:1101:25681:1407
```
definitely seems to be something very wrong here.

**Comment by glennhickey** [comment]
OK, finally had a minute to look into this a little bit more

1st thing I found: the Y-axis on mapping speed supplementary figure in the minigraph-cactus paper is completely wrong.  Due to a awk typo, it's somehow off by a factor of (15/6)!?.  The values should be `732.032, 763.941, 825.612, 948.396, 975.728,973.597` instead of `1830.08,1909.85,2064.03,2370.99,2439.32,2433.99`. I'll update the biorxiv version and make sure this correction gets put into the published version. @hannespetur thanks for raising this issue, I never would have caught this otherwise!! 

Next I tried to re-map the first 500k of these reads (hg003 30x giab) on a r5.8xlarge machine (32 core / 256Gb ram).  I put all the data including the reads and output and logs here: http://public.gi.ucsc.edu/~hickey/debug/slow-giraffe/

I tried two commits (built locally with `make static`):  The commit i used in the paper https://github.com/vgteam/vg/commit/34a822dae4e1cefc01058ccb3887de9183c834b7 from august 10, 2022 and the current master https://github.com/vgteam/vg/commit/1fea4d389392bf56a7992dcbb08ba536b904c8e2 with four graphs from the hprc pages:
[hprc-v1.0-mc-grch38.xg](https://s3-us-west-2.amazonaws.com/human-pangenomics/pangenomes/freeze/freeze1/minigraph-cactus/hprc-v1.0-mc-grch38.xg)
[hprc-v1.0-mc-grch38-minaf.0.1.xg](https://s3-us-west-2.amazonaws.com/human-pangenomics/pangenomes/freeze/freeze1/minigraph-cactus/filtered/hprc-v1.0-mc-grch38-minaf.0.1.xg)
[hprc-v1.0-mc-chm13.xg](https://s3-us-west-2.amazonaws.com/human-pangenomics/pangenomes/freeze/freeze1/minigraph-cactus/hprc-v1.0-mc-chm13.xg)
[hprc-v1.0-mc-chm13-minaf.0.1.xg](https://s3-us-west-2.amazonaws.com/human-pangenomics/pangenomes/freeze/freeze1/minigraph-cactus/filtered/hprc-v1.0-mc-chm13-minaf.0.1.xg)
When running the older commit, I used the ".min.old" and ".dist.old" versions of the indexes from the hprc site. 

The reads / second / thread per giraffe's logs are as follows
```
graph                                  34a822d              1fea4d3
---------------------------------------------------------------------------
hprc-v1.0-mc-grch38                     605.913             <2*
hprc-v1.0-mc-grch38-minaf-0.1           1474.17             1471.16                 
hprc-v1.0-mc-chm13                      707.574             137.868
hprc-v1.0-mc-chm13-minaf-0.1            1549.03             1465.71

* estimate after letting it run for 2 hours without finishing
``` 

On this data set, the allele filtered graphs were twice as fast with the old commit (which is a bigger difference than my figure, but maybe due to the much smaller data set and/or different hardware).

But it also shows a massive regression in performance on the unfiltered graphs.  The grch38-based graph went from 600 reads/sec to totally unmappable.  And the chm13-based graph is about 5X slower. 

I realize that we recommend mapping to the filtered graphs for best performance with DeepVariant. But I am concerned that, over the past year, this has somehow become absolutely mandatory for any mapping experiment.  It's also not a good look that hprc-v1.0-mc-grch38 , which is the main focus of the main paper, has become completely unusable with giraffe.

**Comment by hannespetur** [comment]
Wow, thank you very much for finding time to investigate this so thoroughly! Your benchmark would make be useful to systematically run on every commit, as a performance/speed regression for giraffe, don't you think?

I think it's best for now that I stick with https://github.com/vgteam/vg/commit/34a822dae4e1cefc01058ccb3887de9183c834b7 in my comparison so I can run giraffe on a comparable graph as well as the recommended AF>10% graph.

**Comment by hannespetur** [comment]
Ok thanks, I will switch to v1.43.0

**Comment by glennhickey** [comment]
OK, I think this has been resolved by #3921 / #3923 (thanks @xchang1!).  

I re-uploaded fixed indexes to the HPRC website (https://github.com/human-pangenomics/hpp_pangenome_resources/pull/13), so now all the graphs should be reasonably mappable with any version of `vg` (with the caveat that the `.old` versions need to be used for `vg` versions < v1.44.0). 

I still think there are some question marks about the relative speeds of some of the different graphs we looked at while debugging this, but those can be a separate issue... 

Thanks again, @hannespetur, and please let us know if you have any more problems.
